### ISSUE TITLE
codemeta.json ignored in monorepos?

### ISSUE THREAD

**Issue by ddeboer** [proycon/codemeta-harvester#12]
For example in https://tools.clariah.nl/network-of-terms, is the [`codemeta.json` in the top-level directory](https://github.com/netwerk-digitaal-erfgoed/network-of-terms/blob/master/codemeta.json) being found and processed? Or is the harvester put off by the `package.json` in the (alphabetically) first package of the monorepo [`network-of-terms-catalog`](https://github.com/netwerk-digitaal-erfgoed/network-of-terms/tree/master/packages)?

Could it be because we changed the repository structurally quite a bit and a cached version is used?

```
[harvester info] Git updating cached clone of https://github.com/netwerk-digitaal-erfgoed/network-of-terms...
```

Full log:

```
(log file starts at Fri Dec 23 03:09:55 UTC 2022)

[harvester info] --> Processing network-of-terms (https://github.com/netwerk-digitaal-erfgoed/network-of-terms) [Fri Dec 23 03:09:55 UTC 2022]

[harvester info] Git updating cached clone of https://github.com/netwerk-digitaal-erfgoed/network-of-terms...

[harvester info] Found release v5.6.1

[harvester info] Using v5.6.1

[harvester info] Git reference: v5.6.1

[harvester info] Scanning directory /tmp/codemeta-harvester.cache/network-of-terms for harvestable resources...

[harvester info] found package.json (NodeJS) for network-of-terms, converting to codemeta

[harvester info] Looking for license....

[harvester info] No license file found

[harvester info] Getting contributors from git...

[harvester info] Extracting last and first commit date from git log....

[harvester info] Date created: 2020-07-30T18:10:48Z+0200, date modified: 2022-01-31T12:37:59Z+0100

[harvester info] Querying Github/GitLab API (https://github.com/netwerk-digitaal-erfgoed/network-of-terms)

[harvester info] Found README.md

[harvester info] Found releaseNotes

[harvester info] Querying Zenodo API for DOI (access token provided)...

[harvester info] Looking for TRL information in README.md...

[harvester info] Looking for repostatus information in README.md...

[harvester info] Looking for continuous integration information in README.md...

[harvester info] Looking for documentation links in README.md...

[harvester info] Falling back to git tag (v5.6.1) if no version number is specified...

[harvester info] Inferring repostatus information from git activity (used only as a fallback if not explicitly provided)...

[harvester info] Inferred repostatus https://www.repostatus.org/#active

[harvester info] Looking for repostatus information in README.md in master branch...

[harvester info] Setting group Netwerk Digitaal Erfgoed (NDE)
```

**Comment by proycon** [comment]
TL;DR: harvester is thrown off by (old?) inconsistent git tags in the repo.

> Is the [codemeta.json in the top-level directory](https://github.com/netwerk-digitaal-erfgoed/network-of-terms/blob/master/codemeta.json) being found and processed? Or is the harvester put off by the package.json in the (alphabetically) first package of the monorepo [network-of-terms-catalog](https://github.com/netwerk-digitaal-erfgoed/network-of-terms/tree/master/packages)?

Yes, the codemeta.json is found and taken first. It always takes precedence over almost everything else (see the list at https://github.com/CLARIAH/tool-discovery/blob/master/CONTRIBUTING.md#q-why-is-one-of-my-metadata-fields-wrong-even-though-i-have-the-correct-value)

But indeed, as your log shows, `codemeta.json` seems missed and `package.json` is used.  This is because the harvester seeks your latest stable release (assuming semantic versioning scheme as per our software requirements). It grabs all the git tags that look like semantic versions, sorts them, and grabs the latest one (``git tag --sort version:refname | grep -E "^v?[0-9]+(\.[0-9])*[A-Za-z]?" | tail -n 1)`` ). In your case `v5.6.1` comes out as the latest version so it takes that  
(commit https://github.com/netwerk-digitaal-erfgoed/network-of-terms/commit/f06a99d). However, that tag doesn't even seem to be in your main tree anymore! The tag situation in your repo is a bit messy, probably things were merged into a monorepo at a later point?

This is a bit of a difficult edge case and we'll have to think of a good solution. Rewriting history and removing the old tags is one, but it may be difficult if local copies already have these tags and push them back. Doing a release with a version > 5.6.1 will also work but may not at all be what you want.

**Comment by ddeboer** [comment]
> However, that tag doesn't even seem to be in your main tree anymore!

That’s right, it comes from another repository (catalog) that was merged into the main one to form a monorepo.

So I agree, this is an edge case.

> This is because the harvester seeks your latest stable release (assuming semantic versioning scheme as per our software requirements).

Make sense. Unfortunately, that won’t work for monorepos that keep different tags for each package (rather than incrementing the version number even for packages that haven’t changed with a commit). For example our (semantically versioned) repository tags include the package name, like `@netwerk-digitaal-erfgoed/network-of-terms-query@2.2.1` (yielding an NPM package with [just the version number 2.2.1](https://www.npmjs.com/package/@netwerk-digitaal-erfgoed/network-of-terms-query)).

Another way to find the latest tag would be:

```
$ git describe --tags --abbrev=0
@netwerk-digitaal-erfgoed/network-of-terms-catalog@6.16.1
```

but the harvester still wouldn’t know how to interpret that.

Perhaps the cleanest solution is for a monorepo like this to be submitted to the Tools Registry as separate tools?

**Comment by proycon** [comment]
> > This is because the harvester seeks your latest stable release (assuming semantic versioning scheme as per our software requirements).
> 
> Make sense. Unfortunately, that won’t work for monorepos that keep
> different tags for each package (rather than incrementing the version
> number even for packages that haven’t changed with a commit). For
> example our (semantically versioned) repository tags include the
> package name, like
> ***@***.******@***.***` (yielding an
> NPM package with [just the version number
> ***@***.***/network-of-terms-query)).

I see what you're doing there yeah.

> Another way to find the latest tag would be:
> 
> ```
> $ git describe --tags --abbrev=0
> @***@***.***
> ```
> 
> but the harvester still wouldn’t know how to interpret that.

Yes, it currently doesn't recognize that tag so will just ignore it. 

Using git describe like that would also assume that the chronologically
latest tag is always the right one and that the release is always in the
main/master tree, that needn't be the case either in other edge cases
(release candidate tags, backports, etc)

> Perhaps the cleanest solution is for a monorepo like this to be submitted to the Tools Registry as separate tools? 

Yes, I agree. But even then the git tag problem would
still persist because they share the same git repo, so we have to do a
bit more so it can parse your tagging scheme. I think we could extend
the harvester so you can configure, for reach tool, a custom 'tagprefix'
to look for in git tags. Then you can set it to
***@***.***/network-of-terms-catalog@' for the catalog
and your existing tagging scheme would work. (I wonder how widespread
such a tagging convention is though, I don't think I've seen it before).


That would also mean you'd have multiple codemeta.jsons in your repo, at
different subpaths (that functionality already exists for the harvester).
That needn't be a problem though.

How does that sound?

**Comment by proycon** [comment]
(github seems to be masking things that look like mail addresses when replying by mail, but you get the idea I assume)

**Comment by proycon** [comment]
(Ability to set a tag prefix is now implemented in e58b8a465c35a01069f96f93ad337ad624b72049, still pending release and deployment)
